Adding typechecks to Column Operations

radeusgd opened the pull request:
### Pull Request Description

Closes #6106 

### Important Notes

<!--
- Mention important elements of the design.
- Mention any notable changes to APIs.
-->

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] The documentation has been updated, if necessary.
- [ ] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [ ] If GUI codebase was changed, the GUI was tested when built using `./run ide build`.

radeusgd committed:
initial work

GregoryTravis commented on distribution/lib/Standard/Table/0.0.0-dev/src/Data/Type/Value_Type.enso:
```diff
@@ -186,23 +186,93 @@ type Value_Type
+       Checks that both types can be compared for ordering and runs the
+       following action or reports an error.
+
+       The second argument may be `Nothing` if the type represents null values.
```
Can you clarify what you mean by "represents null values"? Does this describe certain types but not others? Or is it about the possibility that `type_1` is Nothing?

radeusgd replied:
It's the latter, I will try to rephrase this to be clearer.

GregoryTravis commented on distribution/lib/Standard/Table/0.0.0-dev/src/Data/Type/Value_Type_Helpers.enso:
```diff
@@ -75,3 +78,95 @@ find_common_type types strict =
+       > Example
+
+         type_helpers.check_argument_type other Value_Type.expect_text <| operation
+    check_argument_type self value type_check ~action =
```
Could you add a type signature here?

GregoryTravis replied:
Also it might be good to mention that `type_check` is an assertion that can throw an error, rather than a predicate.

radeusgd committed:
Booleans are no longer _automatically_ coerced with numbers

Booleans are no longer _automatically_ coerced with numbers. Adding type unification checks to `iif`, `fill_nothing` and `coalesce`.

radeusgd committed:
fixes

radeusgd committed:
fixing tests

radeusgd committed:
fixes, cr

jdunkerley commented on distribution/lib/Standard/Database/0.0.0-dev/src/Internal/Database_Type_Helpers.enso:
```diff
@@ -0,0 +1,11 @@
+
+import project.Data.Column.Column
+
+## PRIVATE
+   Type helpers for database tables.
+type_helpers =
+    is_column value = value.is_a Column
+    Operation_Type_Helpers.Instance is_column
```
No idea what this is for could we have a little more detail on the doc string please.

jdunkerley commented on distribution/lib/Standard/Table/0.0.0-dev/src/Errors.enso:
```diff
@@ -417,7 +417,7 @@ type Invalid_Value_Type
         prefix = case self.related_column of
             Nothing -> "Expected "
-            column_name -> "Expected " + column_name + " column to have "
+            column_name : Text -> "Expected " + column_name + " column to have "
```
Can we add a catch all here please. I know we shouldn't get here but throwing an error in the `to_display_text` will mean we just can't see anything in IDE if something set wrongly. (e.g. we passed a Column instead of the name)

jdunkerley commented on distribution/lib/Standard/Table/0.0.0-dev/src/Errors.enso:
```diff
@@ -451,16 +451,26 @@ type Column_Type_Mismatch
-        "No common type could have been found for the columns corresponding to ["+self.column_name+"]. If you want to allow mixed types, please retype the columns to the `Mixed` before the concatenation (note however that most Database backends do not support `Mixed` types, so it may work only for the in-memory backend)."
+        types = self.types.map .to_display_text . join ", "
+        prefix = "No common type could have been found for the types: "+types
+        infix = case self.related_column_name of
+            Nothing -> "."
+            column_name : Text -> " when unifying the column ["+column_name+"]."
```
as above.

jdunkerley commented on distribution/lib/Standard/Table/0.0.0-dev/src/Errors.enso:
```diff
@@ -451,16 +451,26 @@ type Column_Type_Mismatch
+       - types: The types that were tried to be unified.
+       - related_column_name: The name of the resulting column that was being
+         unified, if applicable.
+    Error (types : Vector Value_Type) (related_column_name : Nothing|Text)
```
should we default `related_column_name` to `Nothing`?

radeusgd replied:
I thought it could be better to keep it required and specify `Nothing` explicitly. Otherwise, with other errors like that I was sometimes _not adding_ the related column, because it was optional and I was simply forgetting it is supposed to be there. Whereas, if it's a required argument it forces to think if we can 'put something' there.

jdunkerley commented on distribution/lib/Standard/Database/0.0.0-dev/src/Internal/Statement_Setter.enso:
```diff
@@ -35,4 +35,10 @@ fill_hole_default stmt i value = case value of
     _ : Text      -> stmt.setString i value
     _ : Date_Time -> JDBCUtils.setZonedDateTime stmt i value
-    _             -> stmt.setObject i value
+    ## Time_Of_Day and Date_Time sometimes work ok, but sometimes are passed as
```
`Date_Time` should be `Date`?

jdunkerley commented on distribution/lib/Standard/Table/0.0.0-dev/src/Data/Type/Value_Type.enso:
```diff
@@ -186,23 +186,97 @@ type Value_Type
+       Specifies if values of the given type can be compared for ordering.
+    has_ordering : Boolean
+    has_ordering self = case self of
+        Value_Type.Integer _ -> True
```
Why not `Value_Type.Byte`?

jdunkerley replied:
and I think `Boolean` has an order both in Enso and in DB (`False < True`)

radeusgd replied:
Byte - sure, I kind of thought of it as a binary format and not a number, but obviously that is wrong - I will amend.

As for Booleans - I remember a discussion long time ago that ordering booleans does not make sense so Enso should not support this. But I think the decision has been changed since and indeed, Booleans are comparable in Enso now. Then I can add them here as well.

jdunkerley commented on test/Table_Tests/src/In_Memory/Table_Spec.enso:
```diff
@@ -193,7 +196,7 @@ spec =
             x = Column.from_vector 'a' [1, Nothing, 2, Nothing]
             y = Column.from_vector 'b' [0, 0, 0, 0]
-            r = x.zip y skip_missing=False x-> y-> if x == 1 then x else y
+            r = x.zip y skip_nothing=False x-> y-> if x == 1 then x else y
```
Good spot. Can we do a quick scan for where we still have `missing` in the code base as we should align I think.

radeusgd replied:
A quick grep for `_missing` yields only stuff like:
- `error_on_missing_columns` - that should probably stay
- `if_missing` in operations like `find` or `get` - that sounds good as it
- one example which should be renamed.

radeusgd replied:
Oh, and the `count_missing` method! I guess that one could be renamed.

jdunkerley reviewed (commented):
Through quite a lot of the files... will finish review shortly but here are a few comments!

jdunkerley commented on distribution/lib/Standard/Database/0.0.0-dev/src/Data/Table.enso:
```diff
@@ -1523,6 +1518,53 @@ type Table
+
+             table.replace_text 1 '"(.*?)"' '($1)' use_regex=True
+    replace_text : Text | Integer | Column_Selector | Vector (Integer | Text | Column_Selector) -> Text -> Text -> Case_Sensitivity -> Boolean -> Boolean -> Problem_Behavior -> Table
+    replace_text self columns=[0] term="" new_text="" case_sensitivity=Case_Sensitivity.Sensitive only_first=False use_regex=False on_problems=Problem_Behavior.Report_Warning =
```
lets remove this as have removed from the in-memory please.

radeusgd replied:
Sure, I added it before it was removed 😅

jdunkerley commented on distribution/lib/Standard/Database/0.0.0-dev/src/Data/Column.enso:
```diff
@@ -1028,35 +1059,20 @@ type Column
     ## PRIVATE
     naming_helpers self = self.connection.dialect.get_naming_helpers
 
+
```
why the double space?

radeusgd replied:
accident

jdunkerley reviewed (approved):
One slight change please.

Looks really good and great range of test coverage.

radeusgd committed:
cr2 + add type tests that got removed by rebase (Table.replace_text superseded by Column.replace)

radeusgd committed:
remove now obsolete workaround, because the bug was fixed in the meantime

radeusgd committed:
improve error handling, simplify type checking

radeusgd committed:
fixes
